expression/builtin_json.go

```diff
@@ -285,237 +340,375 @@ type jsonTypeFunctionClass struct {
+	args[0].GetType().Decimal = castJSONDirectly
+	sig := &builtinJSONTypeSig{baseStringBuiltinFunc{bf}}
 	return sig.setSelf(sig), nil
 }
 
-func (b *builtinJSONTypeSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinJSONTypeSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
+	var j json.JSON
+	j, isNull, err = b.args[0].EvalJSON(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", isNull, errors.Trace(err)
 	}
-	return JSONType(args, b.ctx.GetSessionVars().StmtCtx)
+	return j.Type(), false, nil
 }
 
 type jsonExtractFunctionClass struct {
 	baseFunctionClass
 }
 
 type builtinJSONExtractSig struct {
-	baseBuiltinFunc
+	baseJSONBuiltinFunc
 }
 
 func (c *jsonExtractFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinJSONExtractSig{newBaseBuiltinFunc(args, ctx)}
+	tps := make([]evalTp, 0, len(args))
+	tps = append(tps, tpJSON)
+	for range args[1:] {
+		tps = append(tps, tpString)
+	}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpJSON, tps...)
+	args[0].GetType().Decimal = castJSONDirectly
+	sig := &builtinJSONExtractSig{baseJSONBuiltinFunc{bf}}
 	return sig.setSelf(sig), nil
 }
 
-func (b *builtinJSONExtractSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinJSONExtractSig) evalJSON(row []types.Datum) (res json.JSON, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	res, isNull, err = b.args[0].EvalJSON(row, sc)
+	if isNull || err != nil {
+		return res, isNull, errors.Trace(err)
 	}
-	return JSONExtract(args, b.ctx.GetSessionVars().StmtCtx)
+	pathExprs := make([]json.PathExpression, 0, len(b.args)-1)
+	for _, arg := range b.args[1:] {
+		var s string
+		s, isNull, err = arg.EvalString(row, sc)
+		if isNull || err != nil {
+			return res, isNull, errors.Trace(err)
+		}
+		pathExpr, err := json.ParseJSONPathExpr(s)
+		if err != nil {
+			return res, true, errors.Trace(err)
+		}
+		pathExprs = append(pathExprs, pathExpr)
+	}
+	var found bool
+	if res, found = res.Extract(pathExprs); !found {
+		return res, true, nil
+	}
+	return res, false, nil
 }
 
 type jsonUnquoteFunctionClass struct {
 	baseFunctionClass
 }
 
 type builtinJSONUnquoteSig struct {
-	baseBuiltinFunc
+	baseStringBuiltinFunc
 }
 
 func (c *jsonUnquoteFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinJSONUnquoteSig{newBaseBuiltinFunc(args, ctx)}
+	bf := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpJSON)
+	bf.tp.Charset, bf.tp.Collate = mysql.DefaultCharset, mysql.DefaultCollationName
```

Comment:
charset and collate has been set in newBaseBuiltinFuncWithTp,
why set them again here?